fix Java Timestamp#usec and Timestamp#at

colinsurprenant opened the pull request:
- fix Java `Timestamp#usec` to correctly return the current time usec and not the epoch usec, fixes #4819
- fix Java `Timestamp#at` to correctly handle `BigDecimal` epoch, relates to #4565

guyboertje commented:
LGTM

Tests fail - but not because of an error in the code rather:
`rspec ./logstash-core/spec/api/lib/api/node_spec.rb:70 # LogStash::Api::Node#hot threads when requesting idle threads should return JIT threads`

colinsurprenant commented:
@guyboertje yeah, I already raised the issue with @ph and @purbon - clearly unrelated.

purbon commented:
@colinsurprenant @guyboertje @ph fix for this unrelated issue can be located at https://github.com/elastic/logstash/pull/4833, feel free to take a look at it.

talevy commented on logstash-core-event-java/src/main/java/com/logstash/Timestamp.java:
```diff
@@ -70,7 +70,9 @@ public String toString() {
     public long usec() {
-        return new Duration(JAN_1_1970.toDateTime(DateTimeZone.UTC), this.time).getMillis();
+        // JodaTime only supports milliseconds precision we can only return usec at millisec precision.
+        // note that getMillis() return milliss since epoch
```
s/milliss/milliseconds

colinsurprenant replied:
yep

talevy commented on logstash-core-event/spec/logstash/timestamp_spec.rb:
```diff
@@ -106,4 +107,50 @@
     end
   end
+
+  context "at" do
```
can we add potentially 3 more tests?

1. arg==nil
2. arg < 0
3. arg != numeric type

colinsurprenant replied:
done!

talevy commented on logstash-core-event/spec/logstash/timestamp_spec.rb:
```diff
@@ -106,4 +107,50 @@
+    it "should try to preserve and report microseconds precision if possible" do
+      # since Java Timestamp relies on JodaTime which supports only milliseconds precision
+      # the usec method will only be precise up to milliseconds.
+      expect(LogStash::Timestamp.at(946702800.123456).usec).to be_within(1000).of(123456)
```
this `usec` test looks like it would be perfect for a randomized test. I'm OK without it

colinsurprenant replied:
I would not change the current specs but also add randomized. let's defer that + create followup issue ok?

talevy replied:
sounds good

talevy commented:
LGTM

colinsurprenant commented:
I squashed and merged into master, 2.x and 2.3. 
the close/merge status and commits history is borked because I squashed, merged into master and cherrypicked into 2.x an 2.3 but forgot to first push new branch squashed history on github. I pushed it after which wreck havoc in the status :(
